File: typescript/src/password.ts
/**
 * Password hashing, verification, and key derivation with safe defaults.
 */

import * as crypto from 'crypto';
import { AlgorithmError, InvalidKeyError, MissingDependencyError } from './errors';

export { AlgorithmError, InvalidKeyError, MissingDependencyError };

const PBKDF2_DEFAULT_ITERATIONS = 100_000;
const SCRYPT_DEFAULT_N = 16384;
const SCRYPT_DEFAULT_R = 8;
const SCRYPT_DEFAULT_P = 1;
const SCRYPT_DEFAULT_DKLEN = 32;
const SCRYPT_DEFAULT_MAXMEM = 64 * 1024 * 1024;
const ARGON2_VERSION = 0x13;

let _bcrypt: any;

try {
  _bcrypt = require('bcrypt');
} catch {
  _bcrypt = null;
}

const _HAS_ARGON2 = typeof (crypto as any).argon2Sync === 'function';
const _SCRYPT_AVAILABLE = typeof (crypto as any).scryptSync === 'function';
const _DEFAULT = _HAS_ARGON2 ? 'argon2id' : (_SCRYPT_AVAILABLE ? 'scrypt' : 'pbkdf2_sha256');

const _ALLOWED = new Set(['argon2id', 'scrypt', 'bcrypt', 'pbkdf2-sha256']);

export interface HashOptions {
  iterations?: number;
  N?: number;
  r?: number;
  p?: number;
  dklen?: number;
  maxmem?: number;
  timeCost?: number;
  memoryCost?: number;
  parallelism?: number;
  rounds?: number;
}

function toBuffer(value: string | Buffer | Uint8Array): Buffer {
  if (Buffer.isBuffer(value)) return value;
  if (value instanceof Uint8Array) return Buffer.from(value);
  return Buffer.from(value, 'utf-8');
}

function phcB64Encode(data: Buffer): string {
  return data.toString('base64').replace(/=+$/g, '');
}

function phcB64Decode(text: string): Buffer {
  const pad = (4 - (text.length % 4)) % 4;
  return Buffer.from(text + '='.repeat(pad), 'base64');
}

function ab64Encode(data: Buffer): string {
  return data.toString('base64').replace(/\+/g, '.').replace(/=+$/g, '');
}

function ab64Decode(text: string): Buffer {
  const s = text.replace(/\./g, '+') + '='.repeat((4 - (text.length % 4)) % 4);
  return Buffer.from(s, 'base64');
}

function normalizeAlgorithm(algorithm?: string): string {
  const name = (algorithm ?? _DEFAULT).toLowerCase().replace(/_/g, '-');
  if (!_ALLOWED.has(name)) {
    throw new AlgorithmError(`Unsupported password algorithm: ${algorithm}`);
  }
  return name;
}

function resolveOptions(algorithm: string, options: HashOptions): Required<HashOptions> {
  const defaults: HashOptions = {
    iterations: PBKDF2_DEFAULT_ITERATIONS,
    N: SCRYPT_DEFAULT_N,
    r: SCRYPT_DEFAULT_R,
    p: SCRYPT_DEFAULT_P,
    dklen: SCRYPT_DEFAULT_DKLEN,
    maxmem: SCRYPT_DEFAULT_MAXMEM,
    timeCost: 3,
    memoryCost: 65536,
    parallelism: 4,
    rounds: 12,
  };

  if (algorithm === 'pbkdf2-sha256' || algorithm === 'argon2id' || algorithm === 'scrypt' || algorithm === 'bcrypt') {
    defaults.dklen = 32;
  }

  return { ...defaults, ...options } as Required<HashOptions>;
}

/**
 * Hash a password. Default algorithm is `argon2id` when the platform provides
 * `crypto.argon2Sync` (Node >= 23.6), otherwise `scrypt`, otherwise
 * `pbkdf2_sha256`. Pass `algorithm` explicitly to pin a specific one.
 */
export function hash(password: string, algorithm?: string, options: HashOptions = {}): string {
  const algo = normalizeAlgorithm(algorithm);
  const opts = resolveOptions(algo, options);

  if (algo === 'argon2id') {
    if (!_HAS_ARGON2) {
      throw new MissingDependencyError(
        'Argon2id requires Node.js crypto.argon2Sync (Node >= 23.6)'
      );
    }
    const salt = crypto.randomBytes(16);
    const derived = (crypto as any).argon2Sync('argon2id', {
      message: Buffer.from(password, 'utf-8'),
      nonce: salt,
      parallelism: opts.parallelism,
      tagLength: opts.dklen,
      memory: opts.memoryCost,
      passes: opts.timeCost,
      version: ARGON2_VERSION,
    }) as Buffer;
    return `$argon2id$v=19$m=${opts.memoryCost},t=${opts.timeCost},p=${opts.parallelism}$${phcB64Encode(salt)}$${phcB64Encode(derived)}`;
  }

  if (algo === 'bcrypt') {
    if (!_bcrypt) {
      throw new MissingDependencyError(
        'bcrypt requires the bcrypt package. Install: npm install bcrypt'
      );
    }
    return _bcrypt.hashSync(password, opts.rounds);
  }

  if (algo === 'scrypt') {
    if (!_SCRYPT_AVAILABLE) {
      throw new MissingDependencyError('scrypt is not available on this platform');
    }
    const salt = crypto.randomBytes(32);
    const N = opts.N;
    const r = opts.r;
    const p = opts.p;
    const dklen = opts.dklen;
    const maxmem = opts.maxmem;
    const ln = Math.log2(N);
    if (Math.floor(ln) !== ln) {
      throw new AlgorithmError('scrypt N must be a power of two');
    }
    const derived = (crypto as any).scryptSync(password, salt, dklen, { N, r, p, maxmem });
    return `$scrypt$ln=${Math.round(ln)},r=${r},p=${p}$${phcB64Encode(salt)}$${phcB64Encode(derived)}`;
  }

  // pbkdf2_sha256
  const iterations = opts.iterations;
  const salt = crypto.randomBytes(32);
  const derived = crypto.pbkdf2Sync(password, salt, iterations, opts.dklen, 'sha256');
  return `$pbkdf2-sha256$${iterations}$${ab64Encode(salt)}$${ab64Encode(derived)}`;
}

export function hash_with(password: string, algorithm: string, options: HashOptions = {}): string {
  return hash(password, algorithm, options);
}

export function verify(password: string, hashed: string): boolean {
  try {
    if (hashed.startsWith('$argon2id$')) {
      if (!_HAS_ARGON2) return false;
      const parts = hashed.split('$');
      if (parts.length !== 6) return false;
      const params: Record<string, string> = {};
      parts[3].split(',').forEach((pair) => {
        const [k, v] = pair.split('=');
        params[k] = v;
      });
      const memory = parseInt(params.m, 10);
      const time = parseInt(params.t, 10);
      const parallelism = parseInt(params.p, 10);
      const salt = phcB64Decode(parts[4]);
      const stored = phcB64Decode(parts[5]);
      if (salt.length === 0 || stored.length === 0) return false;
      const derived = (crypto as any).argon2Sync('argon2id', {
        message: Buffer.from(password, 'utf-8'),
        nonce: salt,
        parallelism,
        tagLength: stored.length,
        memory,
        passes: time,
        version: ARGON2_VERSION,
      }) as Buffer;
      return crypto.timingSafeEqual(derived, stored);
    }

    if (hashed.startsWith('$2a$') || hashed.startsWith('$2b$') || hashed.startsWith('$2y$')) {
      if (!_bcrypt) return false;
      return _bcrypt.compareSync(password, hashed);
    }

    if (hashed.startsWith('$scrypt$')) {
      if (!_SCRYPT_AVAILABLE) return false;
      const parts = hashed.split('$');
      if (parts.length !== 5) return false;
      const params: Record<string, string> = {};
      parts[2].split(',').forEach((pair) => {
        const [k, v] = pair.split('=');
        params[k] = v;
      });
      const ln = parseInt(params.ln, 10);
      const r = parseInt(params.r, 10);
      const p = parseInt(params.p, 10);
      const salt = phcB64Decode(parts[3]);
      const stored = phcB64Decode(parts[4]);
      if (salt.length === 0 || stored.length === 0) return false;
      const derived = (crypto as any).scryptSync(
        password,
        salt,
        stored.length,
        { N: 2 ** ln, r, p, maxmem: 0 }
      );
      return crypto.timingSafeEqual(derived, stored);
    }

    if (hashed.startsWith('$pbkdf2-sha256$')) {
      const parts = hashed.split('$');
      if (parts.length !== 5) return false;
      const iterations = parseInt(parts[2], 10);
      const salt = ab64Decode(parts[3]);
      const stored = ab64Decode(parts[4]);
      if (salt.length === 0 || stored.length === 0) return false;
      const derived = crypto.pbkdf2Sync(password, salt, iterations, stored.length, 'sha256');
      return crypto.timingSafeEqual(derived, stored);
    }
  } catch {
    return false;
  }
  return false;
}

export function derive(
  passphrase: string | Buffer | Uint8Array,
  salt: Buffer,
  length: number,
  algorithm = 'pbkdf2_sha256',
  options: HashOptions = {}
): Buffer {
  if (!salt || salt.length === 0) {
    throw new InvalidKeyError('A non-empty salt is required for key derivation');
  }
  if (length < 1) {
    throw new InvalidKeyError('length must be a positive integer');
  }

  const p = toBuffer(passphrase);
  const algo = normalizeAlgorithm(algorithm);
  const opts = resolveOptions(algo, options);

  if (algo === 'pbkdf2-sha256') {
    return crypto.pbkdf2Sync(p, salt, opts.iterations, length, 'sha256');
  }

  if (algo === 'scrypt') {
    if (!_SCRYPT_AVAILABLE) {
      throw new MissingDependencyError('scrypt is not available on this platform');
    }
    return (crypto as any).scryptSync(p, salt, length, {
      N: opts.N,
      r: opts.r,
      p: opts.p,
      maxmem: opts.maxmem,
    });
  }

  if (algo === 'argon2id') {
    if (!_HAS_ARGON2) {
      throw new MissingDependencyError(
        'Argon2id requires Node.js crypto.argon2Sync (Node >= 23.6)'
      );
    }
    return (crypto as any).argon2Sync('argon2id', {
      message: p,
      nonce: salt,
      parallelism: opts.parallelism,
      tagLength: length,
      memory: opts.memoryCost,
      passes: opts.timeCost,
      version: ARGON2_VERSION,
    }) as Buffer;
  }

  throw new AlgorithmError(`Unsupported key derivation algorithm: ${algorithm}`);
}

File: rust/src/password.rs
//! Password hashing, verification, and key derivation with safe defaults.

use std::str::FromStr;

use argon2::{Algorithm as Argon2Algorithm, Argon2, Params as Argon2Params, Version};
use base64::{engine::general_purpose::STANDARD, engine::general_purpose::STANDARD_NO_PAD, Engine as _};
use pbkdf2::pbkdf2_hmac;
use rand::rngs::OsRng;
use rand::RngCore;
use scrypt::{scrypt, Params as ScryptParams};
use sha2::Sha256;
use subtle::ConstantTimeEq;

use crate::error::{Error, Result};

/// Options for password hashing and key derivation.
#[derive(Clone, Debug)]
pub struct HashOptions {
    /// PBKDF2 iteration count.
    pub iterations: u32,
    /// Argon2 time cost.
    pub time_cost: u32,
    /// Argon2 memory cost in KiB.
    pub memory_cost: u32,
    /// Argon2 parallelism.
    pub parallelism: u32,
    /// Derived key length in bytes.
    pub output_length: usize,
    /// scrypt log₂(N).
    pub log_n: u8,
    /// scrypt block size parameter.
    pub r: u32,
    /// scrypt parallelization parameter.
    pub p: u32,
    /// bcrypt cost factor.
    pub cost: u32,
}

impl Default for HashOptions {
    fn default() -> Self {
        Self {
            iterations: 100_000,
            time_cost: 3,
            memory_cost: 64 * 1024,
            parallelism: 4,
            output_length: 32,
            log_n: 14,
            r: 8,
            p: 1,
            cost: 12,
        }
    }
}

impl HashOptions {
    /// Validates that the cost parameters are internally consistent and
    /// above safe minimums. Called automatically by [`hash_with`] and
    /// [`derive`]; also usable standalone to check a hand-built struct.
    pub fn validate(&self) -> Result<()> {
        if self.iterations == 0 {
            return Err(Error::InvalidParameter(
                "iterations must be at least 1".to_string(),
            ));
        }
        if self.time_cost == 0 {
            return Err(Error::InvalidParameter(
                "argon2 time_cost must be at least 1".to_string(),
            ));
        }
        if self.parallelism == 0 {
            return Err(Error::InvalidParameter(
                "argon2 parallelism must be at least 1".to_string(),
            ));
        }
        // Argon2 requires memory >= 8 * parallelism (RFC 9106).
        if u64::from(self.memory_cost) < 8 * u64::from(self.parallelism) {
            return Err(Error::InvalidParameter(
                "argon2 memory_cost is too low for the requested parallelism".to_string(),
            ));
        }
        // scrypt requires 1 <= log_n < 64 and n = 2^log_n < 2^(128*r/8).
        if self.log_n == 0 || self.log_n >= 64 {
            return Err(Error::InvalidParameter(
                "scrypt log_n must be between 1 and 63".to_string(),
            ));
        }
        if self.r == 0 {
            return Err(Error::InvalidParameter(
                "scrypt r must be at least 1".to_string(),
            ));
        }
        if self.p == 0 {
            return Err(Error::InvalidParameter(
                "scrypt p must be at least 1".to_string(),
            ));
        }
        if self.output_length == 0 {
            return Err(Error::InvalidParameter(
                "output_length must be at least 1".to_string(),
            ));
        }
        // bcrypt accepts cost factors in [4, 31].
        if !(4..=31).contains(&self.cost) {
            return Err(Error::InvalidParameter(
                "bcrypt cost must be between 4 and 31".to_string(),
            ));
        }
        Ok(())
    }
}

fn opts_or_default(options: Option<&HashOptions>) -> HashOptions {
    let mut defaults = HashOptions::default();
    if let Some(o) = options {
        if o.iterations != 0 {
            defaults.iterations = o.iterations;
        }
        if o.time_cost != 0 {
            defaults.time_cost = o.time_cost;
        }
        if o.memory_cost != 0 {
            defaults.memory_cost = o.memory_cost;
        }
        if o.parallelism != 0 {
            defaults.parallelism = o.parallelism;
        }
        if o.output_length != 0 {
            defaults.output_length = o.output_length;
        }
        if o.log_n != 0 {
            defaults.log_n = o.log_n;
        }
        if o.r != 0 {
            defaults.r = o.r;
        }
        if o.p != 0 {
            defaults.p = o.p;
        }
        if o.cost != 0 {
            defaults.cost = o.cost;
        }
    }
    defaults
}

fn normalize_algorithm(name: &str) -> String {
    name.to_lowercase().replace('_', "-")
}

/// Hash a password with the default algorithm (Argon2id).
pub fn hash(password: &str) -> Result<String> {
    hash_with(password, "argon2id", None)
}

/// Hash a password with a chosen algorithm and optional cost parameters.
pub fn hash_with(password: &str, algorithm: &str, options: Option<&HashOptions>) -> Result<String> {
    let algorithm = normalize_algorithm(algorithm);
    let opts = opts_or_default(options);
    opts.validate()?;
    match algorithm.as_str() {
        "argon2id" => hash_argon2id(password, &opts),
        "scrypt" => hash_scrypt(password, &opts),
        "pbkdf2-sha256" => hash_pbkdf2(password, &opts),
        "bcrypt" => hash_bcrypt(password, &opts),
        _ => Err(Error::UnsupportedAlgorithm(algorithm)),
    }
}

fn random_salt(len: usize) -> Vec<u8> {
    let mut salt = vec![0u8; len];
    OsRng.fill_bytes(&mut salt);
    salt
}

fn phc_b64(data: &[u8]) -> String {
    STANDARD_NO_PAD.encode(data)
}

fn phc_b64_decode(text: &str) -> Result<Vec<u8>> {
    STANDARD_NO_PAD
        .decode(text)
        .map_err(|e| Error::Base64(e.to_string()))
}

fn ab64_encode(data: &[u8]) -> String {
    STANDARD
        .encode(data)
        .replace('+', ".")
        .trim_end_matches('=')
        .to_string()
}

fn ab64_decode(text: &str) -> Result<Vec<u8>> {
    let mut s = text.replace('.', "+");
    let pad = (4 - (s.len() % 4)) % 4;
    s.push_str(&"=".repeat(pad));
    STANDARD
        .decode(s)
        .map_err(|e| Error::Base64(e.to_string()))
}

fn hash_argon2id(password: &str, opts: &HashOptions) -> Result<String> {
    let salt = random_salt(16);
    let params = Argon2Params::new(
        opts.memory_cost,
        opts.time_cost,
        opts.parallelism,
        Some(opts.output_length),
    )
    .map_err(|e| Error::InvalidParameter(e.to_string()))?;
    let argon2 = Argon2::new(Argon2Algorithm::Argon2id, Version::V0x13, params);
    let mut out = vec![0u8; opts.output_length];
    argon2
        .hash_password_into(password.as_bytes(), &salt, &mut out)
        .map_err(|e| Error::Internal(e.to_string()))?;
    Ok(format!(
        "$argon2id$v=19$m={},t={},p={}${}${}",
        opts.memory_cost,
        opts.time_cost,
        opts.parallelism,
        phc_b64(&salt),
        phc_b64(&out)
    ))
}

fn hash_scrypt(password: &str, opts: &HashOptions) -> Result<String> {
    let salt = random_salt(32);
    let params = ScryptParams::new(opts.log_n, opts.r, opts.p, opts.output_length)
        .map_err(|e| Error::InvalidParameter(e.to_string()))?;
    let mut out = vec![0u8; opts.output_length];
    scrypt(password.as_bytes(), &salt, &params, &mut out)
        .map_err(|e| Error::Internal(e.to_string()))?;
    Ok(format!(
        "$scrypt$ln={},r={},p={}${}${}",
        opts.log_n,
        opts.r,
        opts.p,
        phc_b64(&salt),
        phc_b64(&out)
    ))
}

fn hash_pbkdf2(password: &str, opts: &HashOptions) -> Result<String> {
    let salt = random_salt(32);
    let mut out = vec![0u8; opts.output_length];
    pbkdf2_hmac::<Sha256>(password.as_bytes(), &salt, opts.iterations, &mut out);
    Ok(format!(
        "$pbkdf2-sha256${}${}${}",
        opts.iterations,
        ab64_encode(&salt),
        ab64_encode(&out)
    ))
}

fn hash_bcrypt(password: &str, opts: &HashOptions) -> Result<String> {
    let cost = if opts.cost == 0 { 12 } else { opts.cost };
    bcrypt::hash(password, cost).map_err(|e| Error::Internal(e.to_string()))
}

/// Verify a password against a stored hash.
pub fn verify(password: &str, hashed: &str) -> Result<bool> {
    if hashed.starts_with("$argon2id$") {
        verify_argon2id(password, hashed)
    } else if hashed.starts_with("$scrypt$") {
        verify_scrypt(password, hashed)
    } else if hashed.starts_with("$pbkdf2-sha256$") {
        verify_pbkdf2(password, hashed)
    } else if hashed.starts_with("$2a$") || hashed.starts_with("$2b$") || hashed.starts_with("$2y$") {
        bcrypt::verify(password, hashed).map_err(|e| Error::Internal(e.to_string()))
    } else {
        Err(Error::InvalidHashFormat(hashed.to_string()))
    }
}

fn verify_argon2id(password: &str, hashed: &str) -> Result<bool> {
    let parts: Vec<&str> = hashed.split('$').collect();
    if parts.len() != 6 {
        return Err(Error::InvalidHashFormat(hashed.to_string()));
    }
    let params = parse_params(parts[3])?;
    let memory = u32::from_str(params.get("m").ok_or_else(|| Error::InvalidHashFormat("m".to_string()))?)
        .map_err(|_| Error::InvalidHashFormat("m".to_string()))?;
    let time = u32::from_str(params.get("t").ok_or_else(|| Error::InvalidHashFormat("t".to_string()))?)
        .map_err(|_| Error::InvalidHashFormat("t".to_string()))?;
    let parallelism = u32::from_str(params.get("p").ok_or_else(|| Error::InvalidHashFormat("p".to_string()))?)
        .map_err(|_| Error::InvalidHashFormat("p".to_string()))?;
    let salt = phc_b64_decode(parts[4])?;
    let stored = phc_b64_decode(parts[5])?;
    if salt.is_empty() || stored.is_empty() {
        return Err(Error::InvalidHashFormat(hashed.to_string()));
    }

    let params = Argon2Params::new(memory, time, parallelism, Some(stored.len()))
        .map_err(|e| Error::InvalidParameter(e.to_string()))?;
    let argon2 = Argon2::new(Argon2Algorithm::Argon2id, Version::V0x13, params);
    let mut out = vec![0u8; stored.len()];
    argon2
        .hash_password_into(password.as_bytes(), &salt, &mut out)
        .map_err(|e| Error::Internal(e.to_string()))?;
    Ok(out.as_slice().ct_eq(stored.as_slice()).into())
}

fn verify_scrypt(password: &str, hashed: &str) -> Result<bool> {
    let parts: Vec<&str> = hashed.split('$').collect();
    if parts.len() != 5 {
        return Err(Error::InvalidHashFormat(hashed.to_string()));
    }
    let params = parse_params(parts[2])?;
    let log_n = u8::from_str(params.get("ln").ok_or_else(|| Error::InvalidHashFormat("ln".to_string()))?)
        .map_err(|_| Error::InvalidHashFormat("ln".to_string()))?;
    let r = u32::from_str(params.get("r").ok_or_else(|| Error::InvalidHashFormat("r".to_string()))?)
        .map_err(|_| Error::InvalidHashFormat("r".to_string()))?;
    let p = u32::from_str(params.get("p").ok_or_else(|| Error::InvalidHashFormat("p".to_string()))?)
        .map_err(|_| Error::InvalidHashFormat("p".to_string()))?;
    let salt = phc_b64_decode(parts[3])?;
    let stored = phc_b64_decode(parts[4])?;
    if salt.is_empty() || stored.is_empty() {
        return Err(Error::InvalidHashFormat(hashed.to_string()));
    }

    let params = ScryptParams::new(log_n, r, p, stored.len())
        .map_err(|e| Error::InvalidParameter(e.to_string()))?;
    let mut out = vec![0u8; stored.len()];
    scrypt(password.as_bytes(), &salt, &params, &mut out)
        .map_err(|e| Error::Internal(e.to_string()))?;
    Ok(out.as_slice().ct_eq(stored.as_slice()).into())
}

fn verify_pbkdf2(password: &str, hashed: &str) -> Result<bool> {
    let parts: Vec<&str> = hashed.split('$').collect();
    if parts.len() != 5 {
        return Err(Error::InvalidHashFormat(hashed.to_string()));
    }
    let iterations = u32::from_str(parts[2])
        .map_err(|_| Error::InvalidHashFormat("iterations".to_string()))?;
    let salt = ab64_decode(parts[3])?;
    let stored = ab64_decode(parts[4])?;
    if salt.is_empty() || stored.is_empty() {
        return Err(Error::InvalidHashFormat(hashed.to_string()));
    }

    let mut out = vec![0u8; stored.len()];
    pbkdf2_hmac::<Sha256>(password.as_bytes(), &salt, iterations, &mut out);
    Ok(out.as_slice().ct_eq(stored.as_slice()).into())
}

fn parse_params(s: &str) -> Result<std::collections::HashMap<String, String>> {
    let mut map = std::collections::HashMap::new();
    for pair in s.split(',') {
        let mut kv = pair.splitn(2, '=');
        let k = kv.next().ok_or_else(|| Error::InvalidHashFormat(s.to_string()))?;
        let v = kv.next().ok_or_else(|| Error::InvalidHashFormat(s.to_string()))?;
        map.insert(k.to_string(), v.to_string());
    }
    Ok(map)
}

/// Derive a key from a passphrase and salt.
pub fn derive(
    passphrase: &str,
    salt: &[u8],
    length: usize,
    algorithm: Option<&str>,
    options: Option<&HashOptions>,
) -> Result<Vec<u8>> {
    if salt.is_empty() {
        return Err(Error::InvalidParameter("salt must not be empty".to_string()));
    }
    if length == 0 {
        return Err(Error::InvalidParameter("length must be positive".to_string()));
    }

    let algorithm = normalize_algorithm(algorithm.unwrap_or("pbkdf2_sha256"));
    let opts = opts_or_default(options);
    opts.validate()?;

    match algorithm.as_str() {
        "pbkdf2-sha256" => {
            let mut out = vec![0u8; length];
            pbkdf2_hmac::<Sha256>(passphrase.as_bytes(), salt, opts.iterations, &mut out);
            Ok(out)
        }
        "scrypt" => {
            let params = ScryptParams::new(opts.log_n, opts.r, opts.p, length)
                .map_err(|e| Error::InvalidParameter(e.to_string()))?;
            let mut out = vec![0u8; length];
            scrypt(passphrase.as_bytes(), salt, &params, &mut out)
                .map_err(|e| Error::Internal(e.to_string()))?;
            Ok(out)
        }
        "argon2id" => {
            let params = Argon2Params::new(
                opts.memory_cost,
                opts.time_cost,
                opts.parallelism,
                Some(length),
            )
            .map_err(|e| Error::InvalidParameter(e.to_string()))?;
            let argon2 = Argon2::new(Argon2Algorithm::Argon2id, Version::V0x13, params);
            let mut out = vec![0u8; length];
            argon2
                .hash_password_into(passphrase.as_bytes(), salt, &mut out)
                .map_err(|e| Error::Internal(e.to_string()))?;
            Ok(out)
        }
        _ => Err(Error::UnsupportedAlgorithm(algorithm)),
    }
}
